habitat-lab/habitat/tasks/rearrange/utils.py

```diff
@@ -457,24 +457,29 @@ def get_robot_spawns(
     start_position = agent.base_pos
 
     state = sim.capture_state()
+    if agent is None:
+        agent = sim.articulated_agent
 
     # Try to place the robot.
     for _ in range(num_spawn_attempts):
-        sim.set_state(state)
+        # Place within `distance_threshold` of the object.
+        start_position = sim.pathfinder.get_random_navigable_point_near(
+            target_position,
+            distance_threshold,
+            island_index=sim.largest_island_idx,
+        )
+        # get_random_navigable_point_near() can return NaNs for start_position.
+        # We want to make sure that the generated start_position is valid
+        if np.isnan(start_position).any():
+            continue
```

Comment:
NOTE: this function is significantly refactored from SIRo. Seems like that code wasn't doing what it claimed to do.
Here we actually want to sample new points from the correct navmesh until we find a pose which is valid (not NaN, within distance threshold, non-colliding).